src/pubsub/ua_pubsub_reader.c

```diff
@@ -1242,50 +1220,67 @@ UA_ReaderGroup_decodeAndProcessRT(UA_Server *server, UA_ReaderGroup *readerGroup
     useMembufAlloc();
 #endif
 
-    /* Decrypt */
 #ifdef UA_ENABLE_PUBSUB_ENCRYPTION
+    UA_Boolean decrypted = false;
+#endif
     UA_NetworkMessage currentNetworkMessage;
     memset(&currentNetworkMessage, 0, sizeof(UA_NetworkMessage));
-
-    size_t payLoadPosition = 0;
-    UA_StatusCode rv =
-        UA_NetworkMessage_decodeHeaders(buf, &payLoadPosition, &currentNetworkMessage);
+    /* Decode headers necessary for checking identifier */
+    UA_StatusCode rv = UA_NetworkMessage_decodeHeadersRT(buf, &currentNetworkMessage);
     if(rv != UA_STATUSCODE_GOOD) {
         UA_LOG_WARNING_READERGROUP(&server->config.logger, readerGroup,
                               "PubSub receive. decoding headers failed");
         goto done;
     }
 
-    rv = verifyAndDecryptNetworkMessage(&server->config.logger, buf, &payLoadPosition,
-                                        &currentNetworkMessage, readerGroup);
-    UA_NetworkMessage_clear(&currentNetworkMessage);
-    if(rv != UA_STATUSCODE_GOOD) {
-        UA_LOG_WARNING_READERGROUP(&server->config.logger, readerGroup,
-                              "Subscribe failed. verify and decrypt network "
-                              "message failed.");
-        goto done;
-    }
-#endif
-
     /* Process for all readers in the ReaderGroup. Each reader might wait for a
      * different PublisherId. So the offsets are different for each. */
     UA_DataSetReader *dsr;
     LIST_FOREACH(dsr, &readerGroup->readers, listEntry) {
-        decodeAndProcessRT(server, readerGroup, dsr, buf);
+        /* Check the message is intended for this reader */
+        rv = UA_DataSetReader_checkIdentifier(server, &currentNetworkMessage, dsr, readerGroup->config);
+        if(rv == UA_STATUSCODE_GOOD) {
+#ifdef UA_ENABLE_PUBSUB_ENCRYPTION
+            if(!decrypted) {
+                UA_NetworkMessage *nm = dsr->bufferedMessage.nm;
+                size_t payLoadPosition = 0;
+                if(!nm) {
+                   /* first call */
+                   nm = &currentNetworkMessage;
+                }
+                rv = UA_NetworkMessage_decodeHeaders(buf, &payLoadPosition, nm);
```

Comment:
If the header is decoded to the buffered message, is the memory for the ext header still allocated in each cycle? Must this memory be freed in between?

Reply:
According to the valgrind build the decoding of the payload header allocates memory which is not freed. Thus UA_NetworkMessage_decodeHeaders can not be called on the buffered network message without freeing the memory in every cycle.